apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,214 @@
+        /// the base consumer.
+        /// </summary>
+        private List<T> _buffer = new List<T>();
+
+        /// <summary>The size of the buffer.</summary>
+        private int _size = 0;
+
+        /// <summary>The number of sequential unsuccessful retries.</summary>
+        private int _retries = 0;        
+
+        internal RpcRetryConsumer(IConsumer<T> consumer, RetryOptions options, 
+            Func<T, int> sizer, Func<Action, RetryOptions, ISequentialThreadingTimer> timerFactory)
+        {
+            _consumer = GaxPreconditions.CheckNotNull(consumer, nameof(consumer));
+            _options = GaxPreconditions.CheckNotNull(options, nameof(options));
+            _sizer = GaxPreconditions.CheckNotNull(sizer, nameof(sizer));
+            _timerFactory = GaxPreconditions.CheckNotNull(timerFactory, nameof(timerFactory));
+
+            // Start the background timer that will retry failed requests.
+            _timer = _timerFactory(FlushBuffer, _options);
+        }
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _consumer.Receive(items);
+            }
+            catch (RpcException e)
+            {
+                HandleException(items, e);
+            }
+        }
+
+        /// <inheritdoc />
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                return _consumer.ReceiveAsync(items, cancellationToken);
+            }
+            catch (RpcException e)
+            {
+                HandleException(items, e);
+                return CommonUtils.CompletedTask;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose() => _timer?.Dispose();
+
+        /// <summary>
+        /// Handles a <see cref="RpcException"/> that occurred during a <see cref="Receive(IEnumerable{T})"/> call.
+        /// </summary>
+        private void HandleException(IEnumerable<T> items, RpcException e)
+        {
+            // If we are not retrying failed calls either propagate or swallow the exception.
+            if (_options.RetryType == RetryType.None)
+            {
+                switch (_options.ExceptionHandling)
+                {
+                    case ExceptionHandling.Propagate:
+                        throw e;
+                    case ExceptionHandling.Swallow:
+                        Debug.WriteLine($"Swallowed exception: {e}");
+                        return;
+                    default:
+                        throw new InvalidOperationException(
+                            $"Invalid ExceptionHandling option: {_options.ExceptionHandling}");
```

Comment:
I'd probably throw this kind of exception when the value is specified when creating options and simply debug assert here (same comment below).